### ISSUE TITLE
aircrack-ng runs into race condition if built with -Wp,-D_FORTIFY_SOURCE=2

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#1701]
**Reported by rebus on 6 Oct 2016 12:10 UTC**

Originally reported in Fedora Bugzilla:
https://bugzilla.redhat.com/show_bug.cgi?id=1348392

Description of problem:
Hello, 
it seems that aircrack-ng runs into some race-condition problems if it is compiled with the -Wp,-D_FORTIFY_SOURCE=2 (which is the default hardening option).

Version-Release number of selected component (if applicable):
aircrack-ng-1.2-0.10rc4

How reproducible:
50%

Steps to Reproduce:
1. rpmbuild -bp aircrack-ng.spec
2. go to the build directory
3. cd src
4. export CFLAGS with -Wp,-D_FORTIFY_SOURCE=2
``` export CFLAGS="-g -W -Wall -O3  -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -mavx2 -DJOHN_AVX2 -pthread -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -DCONFIG_LIBNL30 -DCONFIG_LIBNL -I/usr/include/libnl3  -fstack-protector-strong -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude -DSIMD_CORE -DHAVE_SQLITE -lsqlite3 -DHAVE_PCRE -lpcre --param=ssp-buffer-size=4" ```

5. make all
6. run a test case several times - sometimes it fails, sometime it works fine

``` ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap ```

7. compare to compiling with FORTIFY_SOURCE=0

```
export CFLAGS="-g -W -Wall -O3  -Werror=format-security -Wp,-D_FORTIFY_SOURCE=0 -mavx2 -DJOHN_AVX2 -pthread -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -DCONFIG_LIBNL30 -DCONFIG_LIBNL -I/usr/include/libnl3  -fstack-protector-strong -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude -DSIMD_CORE -DHAVE_SQLITE -lsqlite3 -DHAVE_PCRE -lpcre --param=ssp-buffer-size=4"
```

8. make
make clean
make

9. run test case again

```
./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap 
```

Actual results:
The aircrack-ng sometimes fails to find a key in case it is compiled with -Wp,-D_FORTIFY_SOURCE=2. The failure is not deterministic and it doesn't look like regular segmentation fault, but rather reports "Passphrase not in dictionary" result, which sounds like everything was working fine, but the password was not found.

Expected results:
If there is a memory leak, which triggers some kind of failure, it should SIGSEGV fault instead of reporting "Not found.". Compiling with FORTIFY_SOURCE=0 it deterministically finds a password in the given testcase.

Additional info:
Example session with the FORTIFY_SOURCE=2:
```
2016-06-21 07:38:38 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
Passphrase not in dictionary
2016-06-21 07:38:58 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:38:59 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
Passphrase not in dictionary
2016-06-21 07:39:00 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:01 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:02 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:03 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
Passphrase not in dictionary
2016-06-21 07:39:04 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:05 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
Passphrase not in dictionary
2016-06-21 07:39:06 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:07 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:08 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:09 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
KEY FOUND! [ 12345678 ]
2016-06-21 07:39:09 +0200$ ./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
Passphrase not in dictionary
```

**Comment by aircrack-ng** [comment]
**Comment by misterx on 6 Oct 2016 12:10 UTC**

Might be linked to #1680.

**Comment by aircrack-ng** [comment]
**Comment by samueloph on 6 Oct 2016 12:10 UTC**

I can confirm it is linked/the same problem, just tested compilation with/withouth fortify.

How about we focus the discussion on #1680?

**Comment by aircrack-ng** [comment]
**Modified by misterx on 6 Oct 2016 12:10 UTC**

**Comment by aircrack-ng** [comment]
**Comment by misterx on 6 Oct 2016 12:10 UTC**

Does https://patch-diff.githubusercontent.com/raw/aircrack-ng/aircrack-ng/pull/57.diff fix the issue?

**Comment by aircrack-ng** [comment]
**Comment by samueloph on 6 Oct 2016 12:10 UTC**

Yes, it does :)

Patched: Passes at least 20 consecutive Harkonen tests.

Non-Patched: Segfault randomly, from 10 checks at least 1 segfaults.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 6 Oct 2016 12:10 UTC**

Fixed in aircrack-ng/aircrack-ng@e63ce86096b980c34f2d85a43bb9c73171213bc1.
